scripts/template_matcher.py

```diff
@@ -59,15 +65,69 @@ def _compute_prediction(self, k, img, kp, des):
         kp: keypoints from scene image,   des: descriptors from scene image
         """
 
-        # TODO: find corresponding points in the input image and the templae image
+        # Find corresponding points in the input image and the template image
         #       put keypoints from template image in template_pts
         #       put corresponding keypoints from input image in img_pts
 
+        # Transform input image so that it matches the template image as well as possible
+        
+        bf = cv2.BFMatcher()
+        matches = bf.knnMatch(self.descs[k], des, k=2)
+
+        #Apply ratio test
+        good = []
+        for m,n in matches:
+            if m.distance < self.good_thresh*n.distance:
+                good.append(m)
 
-        #TODO: change img to img_T once you do the homography transform
-        visual_diff = compare_images(img, self.signs[k])
+        if len(good) > 0:
+            template_pts = numpy.float32([ kp[m.trainIdx].pt for m in good ]).reshape(-1,1,2)
+            img_pts = numpy.float32([ self.kps[k][m.queryIdx].pt for m in good ]).reshape(-1,1,2)
+            
+                
+        M, mask = cv2.findHomography(img_pts, template_pts, cv2.RANSAC, 5.0) #the threshold is a guess for now
+        img_T = cv2.warpPerspective(img, M, self.signs[k].shape[::-1])
+
+        visual_diff = compare_images(img_T, self.signs[k])
         return visual_diff
+
 # end of TemplateMatcher class
 
 def compare_images(img1, img2):
-    return 0
\ No newline at end of file
+    mean1 = numpy.average(img1)
+    mean2 = numpy.average(img2)
+    std1 = numpy.std(img1)
+    std2 = numpy.std(img2)
+    
+    if std1 == 0:
+        std1 = 0.01
+    if std2 == 0:
+        std2 = 0.01
+        
+    norm1 = [(elt - mean1)/std1 for elt in img1]
```

Comment:
With numpy you can actually do this operation without a for loop / list comprehension.  This would do the same thing (because of numpy array broadcasting)

```python
norm1 = (img1 - mean1) / std1
```

more on broadcasting:
https://docs.scipy.org/doc/numpy-1.10.1/user/basics.broadcasting.html